crates/codegen/syntax_templates/src/shared/cursor.rs

```diff
@@ -0,0 +1,342 @@
+use std::rc::Rc;
+
+use super::{
+    cst::{Node, RuleNode, TokenNode},
+    kinds::{RuleKind, TokenKind},
+    text_index::{TextIndex, TextRange},
+};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct CursorPathElement {
+    rule_node: Rc<RuleNode>,
+    child_number: usize,
+    text_offset: TextIndex,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct CursorLeaf {
+    node: Node,
+    child_number: usize,
+    text_offset: TextIndex,
+}
+
+impl CursorLeaf {
+    pub fn text_range(&self) -> TextRange {
+        let start = self.text_offset;
+        let end = start + self.node.text_len();
+        start..end
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Cursor {
+    path: Vec<CursorPathElement>,
+    leaf: Option<CursorLeaf>,
+}
+
+impl Cursor {
+    pub(crate) fn new(node: Node) -> Self {
+        Self {
+            path: vec![],
+            leaf: Some(CursorLeaf {
+                node,
+                child_number: 0,
+                text_offset: Default::default(),
+            }),
+        }
+    }
+
+    // Unlike clone, this re-roots at the current node.
+    // It does preserve the correct text offset however,
+    // even though the path is reset.
+    pub fn spawn(&self) -> Self {
+        Self {
+            path: vec![],
+            leaf: self.leaf.clone(),
+        }
+    }
+
+    pub fn is_finished(&self) -> bool {
+        self.leaf.is_none()
+    }
+
+    pub fn node_or_none(&self) -> Option<Node> {
+        self.leaf.as_ref().map(|leaf| leaf.node.clone())
+    }
+
+    pub fn node(&self) -> Node {
+        self.leaf
+            .as_ref()
+            .expect("This cursor is finished, so it has no node.")
+            .node
+            .clone()
+    }
+
+    pub fn text_offset(&self) -> TextIndex {
+        self.leaf
+            .as_ref()
+            .expect("This cursor is finished, so it has no text_offset`")
+            .text_offset
+    }
+
+    pub fn text_range(&self) -> TextRange {
+        self.leaf
+            .as_ref()
+            .expect("This cursor is finished, so it hasn no text_range")
+            .text_range()
+    }
+
+    pub fn path_rule_nodes(&self) -> Vec<Rc<RuleNode>> {
```

Comment:
not sure I understand the use case for `path_rule_nodes()`. I see it is currently unused?

Reply:
So that you can access the parent `RuleNode`s when using a cursor to navigate the tree.